include/clock_control.h

```diff
@@ -72,6 +108,64 @@ static inline int clock_control_off(struct device *dev,
 	return api->off(dev, sys);
 }
 
+/**
+ * @brief Request clock to start with notification when clock has been started.
+ *
+ * User can request delayed start by providing exact information when clock
+ * should be ready. Driver ensures that clock is ready before requested time.
+ * It is the driver responsibility to take into account clock startup time.
+ * When clock is already running user callback will be called from the context
+ * of the function call else it is called from other context (e.g. clock
+ * interrupt).
+ *
+ * @param dev 	  Device.
+ * @param sys	  A pointer to an opaque data representing the sub-system.
+ * @param data	  Data structure containing a callback that is called when
+ *		  action is performed. MUST NOT BE ON STACK!. Can be NULL.
+ * @param timeout When (given in k_cycle32_get()) when clock must be ready.
+ *		  Requested time must be within half of 32 bit span. Bigger
```

Comment:
Would use different name than timeout, which suggests giving up after the timeout.
BTW. In which half?

Reply:
i mentioned once `deadline` instead of timeout (https://github.com/zephyrproject-rtos/zephyr/pull/16675#discussion_r291560845). @tbursztyka what do you think? To me deadline sounds better since as a user i want clock to be running at latest at that moment.

Reply:
>  In which half?

I've clarified it. Please take another look.